presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/Murmur3.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.orc.metadata.statistics;
+
+/**
+ * Murmur3 is successor to Murmur2 fast non-cryptographic hash algorithms.
+ * <p>
+ * Murmur3 32 and 128 bit variants.
+ * 32-bit Java port of https://code.google.com/p/smhasher/source/browse/trunk/MurmurHash3.cpp#94
+ * 128-bit Java port of https://code.google.com/p/smhasher/source/browse/trunk/MurmurHash3.cpp#255
+ * <p>
+ * This is a public domain code with no copyrights.
+ * From homepage of MurmurHash (https://code.google.com/p/smhasher/),
+ * "All MurmurHash versions are public domain software, and the author disclaims all copyright
+ * to their code."
+ * <p>
+ * This class was forked from {@code org.apache.orc.util.Murmur3}.
+ */
+public final class Murmur3
+{
```

Comment:
Why not use the one from Slice? Not sure what this is used for, but it’s not a fast implementation

Reply:
The Slice one uses a different finalization so the results are different. There doesn't seem to be a specification for how to do this (nor does the C++ version have it).

Reply:
> There doesn't seem to be a specification for how to do this (nor does the C++ version have it).

The C++ version *is* the specification. This is how the finalization works in that implementation:

```
  h1 ^= len; h2 ^= len;

  h1 += h2;
  h2 += h1;

  h1 = fmix64(h1);
  h2 = fmix64(h2);

  h1 += h2;
  h2 += h1;
```

It's the same as the version in Slice (https://github.com/airlift/slice/blob/master/src/main/java/io/airlift/slice/Murmur3Hash128.java#L119) and the one in Guava ( https://github.com/google/guava/blob/master/guava/src/com/google/common/hash/Murmur3_128HashFunction.java#L170)

I'm surprised they decided to change it and still call it Murmur3, since it's clearly not.

Reply:
Thanks, I missed that. We could add a “Hive” variant to Slice if performance of this becomes an issue (probably would be if we add write support ... read performance doesn’t matter since it happens infrequently for push down).

Reply:
The biggest issue is how they assemble the long they read out of the byte array. We could improve that locally if necessary. I would prefer not to add a hive variant to Slice, since it's way too specific and has no good reason to exist, really.

Reply:
The C++ code doesn't define how to produce a 64-bit result. The C++ and Guava versions are returning the full 128-bit result.

However, I read the Hive code again, and they are actually quite different, as it's actually a 64-bit hash. The loop only does 8 bytes at a time and is thus computing the same computation for each long, rather than every other long like the 128-bit version.

Reply:
I added a commit to switch the code to `ByteArrays.getLong()` from Slice which uses `Unsafe` to read the value directly.